File: leetcode-python-go/895. 最大频率栈/main.go
package main

import (
	"fmt"
)

type FreqStack struct {
	freq    map[int]int
	group   map[int][]int
	maxFreq int
}

func Constructor() FreqStack {
	return FreqStack{map[int]int{}, map[int][]int{}, 0}
}

func (f *FreqStack) Push(val int) {
	f.freq[val]++
	f.group[f.freq[val]] = append(f.group[f.freq[val]], val)
	f.maxFreq = max(f.maxFreq, f.freq[val])
}

func (f *FreqStack) Pop() int {
	val := f.group[f.maxFreq][len(f.group[f.maxFreq])-1]
	f.group[f.maxFreq] = f.group[f.maxFreq][:len(f.group[f.maxFreq])-1]
	f.freq[val]--
	if len(f.group[f.maxFreq]) == 0 {
		f.maxFreq--
	}
	return val
}

func max(a, b int) int {
	if b > a {
		return b
	}
	return a
}

func main() {
	fmt.Println("test done")
}

File: leetcode-python-go/895. 最大频率栈/main.py
from collections import defaultdict

input = ["FreqStack","push","push","push","push","push","push","pop","pop","pop","pop"],
[[],[5],[7],[5],[7],[4],[5],[],[],[],[]]

class FreqStack:
    def __init__(self):
        self.freq = defaultdict(int)
        self.group = defaultdict(list)
        self.maxFreq = 0

    def push(self, val: int) -> None:
        self.freq[val] += 1
        self.group[self.freq[val]].append(val)
        self.maxFreq = max(self.maxFreq, self.freq[val])

    def pop(self) -> int:
        val = self.group[self.maxFreq].pop()
        self.freq[val] -= 1
        if len(self.group[self.maxFreq]) == 0:
            self.maxFreq -= 1
        return val

FreqStack = FreqStack()
result = []
result.append(FreqStack.push(5))
result.append(FreqStack.push(7))
result.append(FreqStack.push(5))
result.append(FreqStack.push(7))
result.append(FreqStack.push(4))
result.append(FreqStack.push(5))
result.append(FreqStack.pop())
result.append(FreqStack.pop())
result.append(FreqStack.pop())
result.append(FreqStack.pop())
print(result)